### ISSUE TITLE
Change order in performAction of EditableText

### ISSUE THREAD

**Issue by b-cancel** [flutter/flutter#20558]
## Code (main.dart)

```
import 'dart:async';

import 'package:flutter/material.dart';

void main() => runApp(
    MaterialApp(home: new MyApp())
);

class MyApp extends StatefulWidget {
  @override
  MyAppState createState() {
    return new MyAppState();
  }
}

class MyAppState extends State<MyApp> {

  final FocusNode emailFocusNode = new FocusNode();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: new FloatingActionButton(
        onPressed: () => focusEmail(),
        child: new Icon(Icons.remove_red_eye),
      ),
      body: Center(
        child: Form(
          child: TextFormField(
            focusNode: emailFocusNode,
            decoration: new InputDecoration(
              labelText: "Email",
              hintText: 'you@swol.com',
            ),
            onFieldSubmitted: (value) => focusEmail(),
          ),
        ),
      ),
    );
  }

  focusEmail() async{
    print("node focused? " + emailFocusNode.toString());
    if(emailFocusNode.hasFocus)
      print("it HAS focus");
    else
      print("it DOES NOT have focus");
    print("focsuing node" + emailFocusNode.toString());
    FocusScope.of(context).requestFocus(emailFocusNode);
    while(emailFocusNode.hasFocus == false) await Future.delayed(new Duration(milliseconds: 100));
    print("node is definatley focused " + emailFocusNode.toString());
  }
}
```

## Steps to Reproduce

1. run the project
2. click the field
3. notice the prints that pop up when pressing the "floating action button" vs pressing the "enter key" on the keyboard

## Logs

I decided to no post the results of "flutter run --verbose" because it was way to many lines...
but feel free to request the output if needed

```
Analyzing firebase_authentication...
No issues found! (ran in 2.0s)
```

```
[√] Flutter (Channel beta, v0.5.1, on Microsoft Windows [Version 10.0.17134.165], locale en-US)
    • Flutter version 0.5.1 at C:\Program Files\flutter
    • Framework revision c7ea3ca377 (3 months ago), 2018-05-29 21:07:33 +0200
    • Engine revision 1ed25ca7b7
    • Dart version 2.0.0-dev.58.0.flutter-f981f09760

[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at C:\Users\bryan\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b02)
    • All Android licenses accepted.

[√] Android Studio (version 3.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 27.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b02)

[!] VS Code, 64-bit edition (version 1.25.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected devices (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 5.1.1 (API 22) (emulator)

! Doctor found issues in 1 category.
```

**Comment by b-cancel** [comment]
A couple extra notes... this has been feature breaking for me so I tried a couple of other things but they are still not working...

Since the Issue seems to be where I am calling the function from I decided that when I click the "enter key" on the keyboard I would simply "schedule" a change

## Test # 1

I had a variable
var scheduledNode = ValueNotifier<FocusNode>(null);
that had a listener that would run the focusField function by passing in scheduledNode.value

then in the scheduleFocus(ValueNotifier<FocusNode> scheduled, FocusNode fn) function I would
set schedduled.value = fn;

this would trigger the functioning listening to changes and then run the focusField function

[Conclusion] same problem persisted

## Test # 2

using the same variable
var scheduledNode = ValueNotifier<FocusNode>(null);

I instead enclosed my form in an Animated builder which would rebuild when a change was detected
then in the build method I would run focusField if it wasnt null

[Conclusion] same problem persisted

## NOTE: this is ".hasFocus not working issue" is happening in other parts of my code as well

**Comment by b-cancel** [comment]
never mind I realized that this is just an issue a preference

as the code in "editable_text.dart" indicates
'''
 @override
  void performAction(TextInputAction action) {
    switch (action) {
      case TextInputAction.done:
        widget.controller.clearComposing();
        widget.focusNode.unfocus();
        if (widget.onSubmitted != null)
          widget.onSubmitted(_value.text);
        break;
      case TextInputAction.newline:
        // Do nothing for a "newline" action: the newline is already inserted.
        break;
    }
  }
'''

first it unfocuses and then it runs the onSubmitted function
although my expectation was for it to do it in the reverse order

**Comment by b-cancel** [comment]
## Feature Request

As mentioned previously the only reason this is happening is because of the order things are executing in the framework so I ask... 

For override function "void performAction(TextInputAction action)" inside the "editable_text.dart" file
Is it possible to convert the "before" into the "After" ? (show below)

I am asking this because I am making a Plug In that has per field Form validation and it refocuses depending on whether or not the field validated.
If it validates then it focuses on the next field, if it doesn't it refocuses on the current field.

For this to work properly in all cases I need to be able to check if the field that I am submitting is focused or not and that has to happen before I unfocus

let me know if you require more details.

<h3>Before</h3>
```
        widget.controller.clearComposing();
        if (widget.onSubmitted != null)
          widget.onSubmitted(_value.text);
        widget.focusNode.unfocus();
```
<h3>After</h3>
```
        widget.controller.clearComposing();
        if (widget.onSubmitted != null)
          widget.onSubmitted(_value.text);
        widget.focusNode.unfocus(); //location changed
```

**Comment by iapicca** [comment]
Hi @b-cancel 
testing the code below with the latest master
<details><summary>code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) => MaterialApp(
        home: const MyHomePage(key: Key('home')),
      );
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  final FocusNode _emailFocusNode = FocusNode();

  bool _shouldRefocus = true;

  @override
  Widget build(BuildContext context) => Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.lightBlueAccent,
        title: Row(
          children: <Widget>[
            Text('onSubmit: '),
            DropdownButton<bool>(
            
            value: _shouldRefocus,
            onChanged: (bool _bool) {
              setState(() {
                _shouldRefocus = _bool;
              });
            },
            items: <DropdownMenuItem<bool>>[
              const DropdownMenuItem<bool>(
                value: true,
                child: Text('should unfocus and refocus'),
              ),
              const DropdownMenuItem<bool>(
                value: false,
                child: Text('should unfocus'),
              ),
            ]
      ),
  
      ],
      ),),
      floatingActionButton: FloatingActionButton(
        onPressed: () => _emailFocusNode.hasFocus ? _emailFocusNode.unfocus() : focus(from: 'fab'),
        child: Icon(Icons.remove_red_eye),
      ),
      body: Center(
        child: Form(
          child: TextFormField(
            focusNode: _emailFocusNode,
            onFieldSubmitted: (String value) => _shouldRefocus ? focus(from: 'onFieldSubmitted') : null,
          ),
        ),
      ),
    );

  Future<void> _keepChecking() async {
    for (int i = 0; i < 10000 &&( !_emailFocusNode.hasFocus); ++i ) {
      await Future.delayed(const Duration(microseconds: 1));
      print('still not focused ($i)');
    }
  }
  
  void focus({@required String from}) async {
    print('[$from] is focused: ${_emailFocusNode.hasFocus}');
    print('focusing...');
    FocusScope.of(context).requestFocus(_emailFocusNode);
    await _keepChecking();
    print('[$from] is focused: ${_emailFocusNode.hasFocus}');
  }
}

```

</details>
<details><summary>logs</summary>

```bash
[        ] Waiting for Pixel 3a to report its views...
[  +11 ms] Waiting for Pixel 3a to report its views... (completed in 9ms)
[  +14 ms] DevFS: Creating new filesystem on the device (null)
[  +36 ms] DevFS: Created new filesystem on the device (file:///data/user/0/com.example.master/code_cache/masterYPLTZE/master/)
[   +3 ms] Updating assets
[ +115 ms] Syncing files to device Pixel 3a...
[   +3 ms] Scanning asset files
[   +3 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +1 ms] <- recompile package:master/main.dart b54ea8b4-d23f-45b6-a5ce-1c99def1c05e
[        ] <- b54ea8b4-d23f-45b6-a5ce-1c99def1c05e
[  +97 ms] Updating files
[ +216 ms] DevFS: Sync finished
[   +1 ms] Syncing files to device Pixel 3a... (completed in 323ms)
[   +2 ms] Synced 0.9MB.
[   +2 ms] <- accept
[   +7 ms] Connected to _flutterView/0x708202d820.
[   +2 ms] Flutter run key commands.
[   +3 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[   +1 ms] q Quit (terminate the application on the device).
[        ] An Observatory debugger and profiler on Pixel 3a is available at: http://127.0.0.1:55714/vxmcaOMpCLQ=/
[+44181 ms] I/flutter (19214): [fab] is focused: false
[        ] I/flutter (19214): focusing...
[  +57 ms] W/IInputConnectionWrapper(19214): getExtractedText on inactive InputConnection
[   +2 ms] W/IInputConnectionWrapper(19214): getTextBeforeCursor on inactive InputConnection
[   +5 ms] W/IInputConnectionWrapper(19214): getExtractedText on inactive InputConnection
[        ] W/IInputConnectionWrapper(19214): getTextBeforeCursor on inactive InputConnection
[   +3 ms] I/flutter (19214): still not focused (0)
[        ] I/flutter (19214): [fab] is focused: true
[  +11 ms] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[  +25 ms] I/chatty  (19214): uid=10364(com.example.master) identical 4 lines
[        ] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[+14603 ms] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[  +10 ms] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[        ] I/flutter (19214): [onFieldSubmitted] is focused: true
[        ] I/flutter (19214): focusing...
[   +3 ms] I/flutter (19214): [onFieldSubmitted] is focused: true
[+8637 ms] W/IInputConnectionWrapper(19214): beginBatchEdit on inactive InputConnection
[        ] W/IInputConnectionWrapper(19214): setComposingRegion on inactive InputConnection
[        ] W/IInputConnectionWrapper(19214): endBatchEdit on inactive InputConnection
[   +6 ms] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[  +23 ms] I/chatty  (19214): uid=10364(com.example.master) identical 2 lines
[        ] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[ +879 ms] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[  +16 ms] E/SpannableStringBuilder(19214): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
[+15337 ms] DevFS: Deleting filesystem on the device (file:///data/user/0/com.example.master/code_cache/masterYPLTZE/master/)
[  +43 ms] DevFS: Deleted filesystem on the device (file:///data/user/0/com.example.master/code_cache/masterYPLTZE/master/)
[ +288 ms] Service protocol connection closed.
[        ] Application finished.
[   +5 ms] executing: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +7 ms] Exit code 0 from: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[        ] 965AY0WP5C tcp:55714 tcp:37477
[   +1 ms] executing: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --remove tcp:55714
[  +12 ms] executing: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +7 ms] Exit code 0 from: /Users/nevercode/Library/Android/sdk/platform-tools/adb -s 965AY0WP5C forward --list
[   +2 ms] "flutter run" took 117,950ms.
[   +5 ms] Running shutdown hooks
[        ] Shutdown hook priority 4
[        ] Shutdown hooks complete
[        ] exiting with code 0
```

</details>
<details><summary>doctor</summary>

```console
[✓] Flutter (Channel master, 1.21.0-2.0.pre.94, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.21.0-2.0.pre.94 at /Users/nevercode/development/flutter_master
    • Framework revision adc9dde3ba (9 hours ago), 2020-07-21 16:17:23 -0700
    • Engine revision 5a73d4dc25
    • Dart version 2.9.0 (build 2.9.0-21.0.dev 84e5eeaea4)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (4 available)
    • Pixel 3a (mobile) • 965AY0WP5C • android-arm64  • Android 10 (API 29)
    • macOS (desktop)   • macos      • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server (web)  • web-server • web-javascript • Flutter Tools
    • Chrome (web)      • chrome     • web-javascript • Google Chrome 84.0.4147.89

• No issues found!
```

</details>

`onSubmit` fails to unfocus while the fab works correctly
Thank you

**Comment by justinmc** [comment]
@iapicca Could you explain what exactly the desired behavior is here?  I'm playing around with the code locally but not understanding everything.

The code in editable_text.dart seems to have changed a bit since the issue was opened.  Maybe this can be worked around now by using onEditingComplete.

**Comment by darshankawar** [comment]
I tried to verify this issue using below code sample (updated to reflect latest version) and then tried to replicate by tapping `enter` key and tapping on FAB which gave me below results in the console:

```
flutter: [fab] is focused: false
flutter: focusing...
flutter: still not focused (0)
flutter: [fab] is focused: true
flutter: [onFieldSubmitted] is focused: true
flutter: focusing...
flutter: [onFieldSubmitted] is focused: true
flutter: [fab] is focused: false
flutter: focusing...
flutter: still not focused (0)
flutter: [fab] is focused: true

```

<details>
<summary> updated runnable code sample </summary>

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => MaterialApp(
    home: const MyHomePage(key: Key('home')),
  );
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final FocusNode _emailFocusNode = FocusNode();

  bool _shouldRefocus = true;

  @override
  Widget build(BuildContext context) => Scaffold(
    appBar: AppBar(
      backgroundColor: Colors.lightBlueAccent,
      title: Row(
        children: <Widget>[
          Text('onSubmit: '),
          DropdownButton<bool>(
              value: _shouldRefocus,
              onChanged: (bool? _bool) {
                setState(() {
                  _shouldRefocus = _bool!;
                });
              },
              items: <DropdownMenuItem<bool>>[
                const DropdownMenuItem<bool>(
                  value: true,
                  child: Text('should unfocus and refocus'),
                ),
                const DropdownMenuItem<bool>(
                  value: false,
                  child: Text('should unfocus'),
                ),
              ]),
        ],
      ),
    ),
    floatingActionButton: FloatingActionButton(
      onPressed: () => _emailFocusNode.hasFocus
          ? _emailFocusNode.unfocus()
          : focus(from: 'fab'),
      child: Icon(Icons.remove_red_eye),
    ),
    body: Center(
      child: Form(
        child: TextFormField(
          focusNode: _emailFocusNode,
          onFieldSubmitted: (String value) =>
          _shouldRefocus ? focus(from: 'onFieldSubmitted') : null,
        ),
      ),
    ),
  );

  Future<void> _keepChecking() async {
    for (int i = 0; i < 10000 && (!_emailFocusNode.hasFocus); ++i) {
      await Future.delayed(const Duration(microseconds: 1));
      print('still not focused ($i)');
    }
  }

  void focus({required String from}) async {
    print('[$from] is focused: ${_emailFocusNode.hasFocus}');
    print('focusing...');
    FocusScope.of(context).requestFocus(_emailFocusNode);
    await _keepChecking();
    print('[$from] is focused: ${_emailFocusNode.hasFocus}');
  }
}

```

</details>

Also looking at the original issue and per @justinmc above comment, I am not completely sure what should be the expected behavior or if this isn't working as intended.

Considering age of the issue and lack of activity after last comment, I am going ahead and closing this as stale.

@b-cancel If you think this is incorrect and think this is still an issue, I suggest you to open a new issue with all relevant details using latest stable version of the framework so that we can properly address this issue.
